packages/core/src/dialects/abstract/query-generator-typescript.ts

```diff
@@ -372,6 +378,143 @@ export class AbstractQueryGeneratorTypeScript {
     throw new Error(`showConstraintsQuery has not been implemented in ${this.dialect.name}.`);
   }
 
+  addIndexQuery(tableName: TableNameOrModel, options: AddIndexQueryOptions): string {
+    const ADD_INDEX_QUERY_SUPPORTED_OPTIONS = new Set<keyof AddIndexQueryOptions>();
+    if (this.dialect.supports.index.concurrently) {
+      ADD_INDEX_QUERY_SUPPORTED_OPTIONS.add('concurrently');
+    }
+
+    if (this.dialect.supports.index.ifNotExists) {
+      ADD_INDEX_QUERY_SUPPORTED_OPTIONS.add('ifNotExists');
+    }
+
+    if (this.dialect.supports.index.type) {
+      ADD_INDEX_QUERY_SUPPORTED_OPTIONS.add('indexType');
+    }
+
+    if (this.dialect.supports.index.include) {
+      ADD_INDEX_QUERY_SUPPORTED_OPTIONS.add('include');
+    }
+
+    if (this.dialect.supports.index.operator) {
+      ADD_INDEX_QUERY_SUPPORTED_OPTIONS.add('operator');
+    }
+
+    if (this.dialect.supports.index.parser) {
+      ADD_INDEX_QUERY_SUPPORTED_OPTIONS.add('parser');
+    }
+
+    if (this.dialect.supports.index.using) {
+      ADD_INDEX_QUERY_SUPPORTED_OPTIONS.add('using');
+    }
+
+    if (this.dialect.supports.index.where) {
+      ADD_INDEX_QUERY_SUPPORTED_OPTIONS.add('where');
+    }
+
+    rejectInvalidOptions(
+      'addIndexQuery',
+      this.dialect.name,
+      ADD_INDEX_QUERY_SUPPORTABLE_OPTIONS,
+      ADD_INDEX_QUERY_SUPPORTED_OPTIONS,
+      { ...options },
+    );
```

Comment:
I think we should rework `rejectInvalidOptions` first to accept `dialect.supports` (or a dialect and derive `name` and `supports` from it) instead of the `SUPPORTED_OPTIONS` set and then move all of the sets (and the features) over to `dialect.supports`. This shows how much duplication we're doing

Reply:
I agree with this.
It's going to take a bit of work to get all the options reconfigured (outside the scope of this PR).

Reply:
Oh definitely outside the scope of this PR